File: 0077-combinations/0077-combinations.java
class Solution {
    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> result = new ArrayList<>();
        findCombinations(1, n, k, new ArrayList<>(), result);
        return result;
    }

    private void findCombinations(int start, int end, int k, List<Integer> path, List<List<Integer>> result) {
        for(int i = start; i <= end; i++) {
            path.add(i);
            int size = path.size();
            if (size == k) {
                result.add(new ArrayList<>(path));
            } else {
                findCombinations(i+1, end, k, path, result);
            }
            path.removeLast();
        }
    }


}

File: 0077-combinations/0077-combinations.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        nums = [i for i in range(1, n+1)]
        result = []
        self.build(0, nums, k, [], result)
        return result


    def build(self, start, nums, k, path, result): 
        if len(path) == k:
            result.append(path[:])
            return 

        for i in range(start, len(nums)): 
            n = nums[i]
            path.append(n)
            self.build(i+1, nums, k, path, result)
            path.pop()
        
